amethyst_assets/src/storage.rs

```diff
@@ -112,46 +122,147 @@ impl<A: Asset> AssetStorage<A> {
+                        // together with a `Data` only once.
+                        unsafe {
+                            assets.insert(id, asset);
+                        }
+
+                        (reload_obj, handle)
+                    }
+                    Processed::HotReload {
+                        data,
+                        format,
+                        handle,
+                        name,
+                        old_reload,
+                    } => {
+                        let (asset, reload_obj) = match data.map(
+                            |FormatValue { data, reload }| (data, reload),
+                        ).and_then(|(d, rel)| f(d).map(|a| (a, rel)))
+                            .map_err(|e| AssetError::new(name, format, e))
+                        {
+                            Ok(x) => x,
+                            Err(e) => {
+                                eprintln!("Failed to hot-reload: {}", e);
+
+                                reloads.push((handle.downgrade(), old_reload));
+
+                                return Ok(());
+                            }
+                        };
+
+                        let id = handle.id();
+                        assert!(bitset.contains(id));
+                        unsafe {
+                            let old = assets.get_mut(id);
+                            *old = asset;
+                        }
+
+                        (reload_obj, handle)
+                    }
+                };
+
+                // Add the reload obj if it is `Some`.
+                if let Some(reload_obj) = reload_obj {
+                    reloads.push((handle.downgrade(), reload_obj));
                 }
 
                 Ok(())
             });
         }
 
-        while let Some(i) = self.handles.iter().position(Handle::is_unused) {
-            let old = self.handles.swap_remove(i);
-            let id = i as u32;
+        let mut skip = 0;
+        while let Some(i) = self.handles.iter().skip(skip).position(Handle::is_unique) {
+            skip = i;
+            let handle = self.handles.swap_remove(i);
+            let id = handle.id();
             unsafe {
                 self.assets.remove(id);
             }
             self.bitset.remove(id);
-            self.unused_handles.push(old);
+
+            // Can't reuse old handle here, because otherwise weak handles would still be valid.
+            // TODO: maybe just store u32?
+            self.unused_handles.push(Handle {
+                id: Arc::new(id),
+                marker: PhantomData,
+            });
+        }
+
+        if strategy
+            .map(|s| s.needs_reload(&mut self.reload_counter))
+            .unwrap_or(false)
+        {
+            self.hot_reload(pool);
+        }
+    }
+
+    fn hot_reload(&mut self, pool: &ThreadPool) {
+        let elapsed = self.last_reload.elapsed().as_secs();
+
+        if elapsed >= 1 {
```

Comment:
You've already checked that reload is required with strategy.

Reply:
Oops, forgot to remove the old code.